ESMCI/cime
# Large refactor: properly encapsulate TestStatus handling

Changes to TestStatus handling:
1) Remove as much as possible TestStatus handling from test_scheduler. Users can run phases by hand (case.build) and therefore it's better for the individual implementations of the phase to update TestStatus.
2) Make a new class TestStatus to encapsulate reading/writing of TestStatus and to provide the valid phases and phase statuses.
3) It is now required that each phase be completely unique. COMPARE phases are decorated with the thing that's being compared (e.g. COMPARE_baseline) to make it unique.
4) The COMMENT system is gone, instead, any tokens that come after the core (STATUS TEST PHASE) are assumed to be comments.

Changes to SystemTests:
1) The base class provides a common framework for the build and run phases. Sub classes may override build_phase or run_phase to define how their phases behave, but they may NOT override the framework (SystemTests.build, SystemTest.run). The idea is to ensure that proper exception handling and TestStatus management occurs no matter what.
2) SystemTestCommon now handles TestStatus throught the TestStatus class

Other:
1) In order to take advantage of automatic code-checking and unit-test discovery, move implementations of all key python tools into utils/python as libraries.
1.a) jenkins_generic_job.py is moved to utils/python because it's an ACME-specific thing, not part of CIME.
1.b) Everything else was moved to utils/python/CIME


Test suite: scripts_regression_tests
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes #330 , #314

User interface changes?: Most important are changes in how to define a system_test. @billsacks please note this for your lii.py test.

Code review: @jedwards4b @billsacks
